### PR TITLE
merge enginevinylsoundemu with pregain and remove dithering

### PR DESCRIPTION
**Pull request by daschuer**
Please try to hear a difference, I cannot. But there are big differences if you check the three scalers.
Even if there is a marginal difference, we may consider if the performance win of on buffer process cycle is worth to merge this PR.

### CODE REVIEW

**Comment by rryan** [comment]
Performance shouldn't motivate this change -- based on profiling (which I've been doing a lot of today!) and optimizations turned off EngineVinylSoundEmu accounts for 0.3% of time spent in callback. Pretty tiny and on the same order of magnitude as a ControlObject::set call.

**Comment by rryan** [comment]
> Performance shouldn't motivate this change -- based on profiling (which I've been doing a lot of today!) and optimizations turned off EngineVinylSoundEmu accounts for 0.3% of time spent in callback. Pretty tiny and on the same order of magnitude as a ControlObject::set call.

Oh, duh -- it's not doing anything since I'm playing the deck at normal speed. Nevermind!

**Comment by ywwg** [comment]
I don't have time to spend time on this issue right now because I don't think it's proven to be broken in any way.  We've had a lot of issues with regressions recently so I want to minimize code changes.

**Comment by daschuer** [comment]
Sorry, but it is. You have pointed it out yourselves. It can lit the clipping light. And since the Gain reduction comes in only below 7% it is a real issue (against my earlier findings)

**Comment by ywwg** [comment]
We have been discussing adding headroom, and with headroom dithering will not add enough to cause clipping.

**Comment by daschuer** [comment]
The current discussion was heading towards a master gain scaling. This will not help to switch of the clipping light of the deck.

File: src/engine/enginepregain.cpp [hunk 1]
```diff
@@ -107,10 +111,17 @@ void EnginePregain::process(CSAMPLE* pInOut, const int iBufferSize) {
     // Clamp gain to within [0, 10.0] to prevent insane gains. This can happen
     // (some corrupt files get really high replay gain values).
     // 10 allows a maximum replay Gain Boost * calculated replay gain of ~2
-    fGain = fGain * math_clamp(fReplayGainCorrection, 0.0f, 10.0f);
+    fGain *= math_clamp(fReplayGainCorrection, 0.0f, 10.0f);
 
     m_pTotalGain->set(fGain);
 
+    // We have to Vinyl emulation near stop, because otherwise we have clicks when
```
Comment (ywwg) [inline src/engine/enginepregain.cpp]: This isn't quite accurate - the problem is not with clicking when reversing direction (the stretching algorithm actually handles this correctly).  Instead, say:

// As the speed approaches zero, hearing small bursts of sound at full volume is distracting and doesn't mimic the way that vinyl sounds when played slowly.  Instead, reduce gain to provide a soft rolloff.

File: src/engine/enginepregain.h [hunk 2]
```diff
@@ -30,9 +30,11 @@ class EnginePregain : public EngineObject {
     EnginePregain(QString group);
     virtual ~EnginePregain();
 
+    void setSpeed(double speed);
     void process(CSAMPLE* pInOut, const int iBufferSize);
 
   private:
+    double m_dSpeed;
```
Comment (ywwg) [inline src/engine/enginepregain.h]: initialize to zero in constructor

**Comment by ywwg** [comment]
I did some testing and couldn't tell the difference with dithering and without.  I'm ok getting rid of it to prevent unnecessary clipping for fully-saturated tracks.  I made a couple notes.

**Comment by daschuer** [comment]
Thank you for your testing and review effort!

**Pull request merged**